### PR TITLE
Build Python 3.13 wheels (not free-threaded)

### LINKED ISSUE #698: 3.13 support
Howdy!

I'm trying to test cattrs on Python 3.13 (beta1). I've added a msgspec converter in the next version so now we optionally depend on msgspec, which means I can't test on 3.13 until you support it. ;)

The installation seems to be failing in GCC: https://github.com/python-attrs/cattrs/actions/runs/9333591881/job/25690767935

```
msgspec/_core.c:11254:15: error: too few arguments to function 
  ‘_PyLong_AsByteArray’
  11254 |     int out = _PyLong_AsByteArray((PyLongObject *)int128, buf, 16, 0, 
  0);
        |               ^~~~~~~~~~~~~~~~~~~
  In file included from 
  /opt/hostedtoolcache/Python/3.13.0-beta.1/x64/include/python3.13/longobject.h:10
  7,
                   from 
  /opt/hostedtoolcache/Python/3.13.0-beta.1/x64/include/python3.13/Python.h:72,
                   from msgspec/_core.c:9:
  /opt/hostedtoolcache/Python/3.13.0-beta.1/x64/include/python3.13/cpython/longobj
  ect.h:99:17: note: declared here
     99 | PyAPI_FUNC(int) _PyLong_AsByteArray(PyLongObject* v,
        |                 ^~~~~~~~~~~~~~~~~~~
  error: command '/usr/bin/gcc' failed with exit code 1
```

### LINKED ISSUE #741: Doesn't install on python 3.13
### Description

OS: macOS Sonoma 14.6.1
Python 3.13. 

```
pip install msgspec
Collecting msgspec
  Downloading msgspec-0.18.6.tar.gz (216 kB)
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
Building wheels for collected packages: msgspec
  Building wheel for msgspec (pyproject.toml) ... error
  error: subprocess-exited-with-error
  
  × Building wheel for msgspec (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> [55 lines of output]
      running bdist_wheel
      running build
      running build_py
      creating build/lib.macosx-10.13-universal2-cpython-313/msgspec
      copying msgspec/msgpack.py -> build/lib.macosx-10.13-universal2-cpython-313/msgspec
      copying msgspec/_version.py -> build/lib.macosx-10.13-universal2-cpython-313/msgspec
      copying msgspec/__init__.py -> build/lib.macosx-10.13-universal2-cpython-313/msgspec
      copying msgspec/yaml.py -> build/lib.macosx-10.13-universal2-cpython-313/msgspec
      copying msgspec/structs.py -> build/lib.macosx-10.13-universal2-cpython-313/msgspec
      copying msgspec/inspect.py -> build/lib.macosx-10.13-universal2-cpython-313/msgspec
      copying msgspec/_json_schema.py -> build/lib.macosx-10.13-universal2-cpython-313/msgspec
      copying msgspec/json.py -> build/lib.macosx-10.13-universal2-cpython-313/msgspec
      copying msgspec/_utils.py -> build/lib.macosx-10.13-universal2-cpython-313/msgspec
      copying msgspec/toml.py -> build/lib.macosx-10.13-universal2-cpython-313/msgspec
      copying msgspec/msgpack.pyi -> build/lib.macosx-10.13-universal2-cpython-313/msgspec
      copying msgspec/__init__.pyi -> build/lib.macosx-10.13-universal2-cpython-313/msgspec
      copying msgspec/json.pyi -> build/lib.macosx-10.13-universal2-cpython-313/msgspec
      copying msgspec/structs.pyi -> build/lib.macosx-10.13-universal2-cpython-313/msgspec
      copying msgspec/py.typed -> build/lib.macosx-10.13-universal2-cpython-313/msgspec
      UPDATING build/lib.macosx-10.13-universal2-cpython-313/msgspec/_version.py
      set build/lib.macosx-10.13-universal2-cpython-313/msgspec/_version.py to '0.18.6'
      running build_ext
      building 'msgspec._core' extension
      creating build/temp.macosx-10.13-universal2-cpython-313/msgspec
      clang -fno-strict-overflow -Wsign-compare -Wunreachable-code -fno-common -dynamic -DNDEBUG -g -O3 -Wall -arch arm64 -arch x86_64 "-I/Users/yangcao/Library/Mobile Documents/com~apple~CloudDocs/Repositories/monorepo/ewx-root/context/ewx-opsy/.venv/include" -I/Library/Frameworks/Python.framework/Versions/3.13/include/python3.13 -c msgspec/_core.c -o build/temp.macosx-10.13-universal2-cpython-313/msgspec/_core.o
      msgspec/_core.c:500:13: error: call to undeclared function '_PyUnicode_EQ'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
        500 |         if (_PyUnicode_EQ(kwname, key)) {
            |             ^
      msgspec/_core.c:4452:12: error: call to undeclared function '_PySet_NextEntry'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
       4452 |     while (_PySet_NextEntry(state->structs_set, &set_pos, &set_item, &set_hash)) {
            |            ^
      msgspec/_core.c:7316:17: error: call to undeclared function '_PyUnicode_EQ'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
       7316 |             if (_PyUnicode_EQ(kwname, field)) {
            |                 ^
      msgspec/_core.c:7723:17: error: call to undeclared function '_PyUnicode_EQ'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
       7723 |             if (_PyUnicode_EQ(kwname, field)) goto kw_found;
            |                 ^
      msgspec/_core.c:11254:72: error: too few arguments to function call, expected 6, have 5
       11254 |     int out = _PyLong_AsByteArray((PyLongObject *)int128, buf, 16, 0, 0);
             |               ~~~~~~~~~~~~~~~~~~~                                      ^
      /Library/Frameworks/Python.framework/Versions/3.13/include/python3.13/cpython/longobject.h:111:17: note: '_PyLong_AsByteArray' declared here
        111 | PyAPI_FUNC(int) _PyLong_AsByteArray(PyLongObject* v,
            |                 ^                   ~~~~~~~~~~~~~~~~
        112 |     unsigned char* bytes, size_t n,
            |     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        113 |     int little_endian, int is_signed, int with_exceptions);
            |     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      msgspec/_core.c:12426:12: error: call to undeclared function '_PySet_NextEntry'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
       12426 |     while (_PySet_NextEntry(obj, &ppos, &item, &hash)) {
             |            ^
      msgspec/_core.c:13748:12: error: call to undeclared function '_PySet_NextEntry'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
       13748 |     while (_PySet_NextEntry(obj, &ppos, &item, &hash)) {
             |            ^
      7 errors generated.
      error: command '/usr/local/opt/llvm/bin/clang' failed with exit code 1
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
  ERROR: Failed building wheel for msgspec
Failed to build msgspec
ERROR: ERROR: Failed to build installable wheels for some pyproject.toml based projects (msgspec)

```

### PR DESCRIPTION
**Pull request by edgarrmondragon**
I can close this if https://github.com/jcrist/msgspec/pull/703 gets more attention, but I've confirmed changes here allow the wheels to be built in 3.8-3.13, and all tests are passing.

This warning is currently raised, but fixing it might be beyond the scope of this PR:

```
PytestUnraisableExceptionWarning: Exception ignored in PyDict_GetItem(); consider using PyDict_GetItemRef() or PyDict_GetItemWithError(): None
```

https://github.com/python/cpython/issues/106004

- Closes #698
- Closes #741

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -99,10 +100,10 @@ jobs:
       - name: Set up Environment
         if: github.event_name != 'release'
         run: |
-            echo "CIBW_SKIP=${CIBW_SKIP} *-musllinux_* cp38-*_aarch64 cp39-*_aarch64 cp311-*_aarch64 cp312-*_aarch64" >> $GITHUB_ENV
+            echo "CIBW_SKIP=${CIBW_SKIP} *-musllinux_* cp38-*_aarch64 cp39-*_aarch64 cp311-*_aarch64 cp312-*_aarch64 cp313-*" >> $GITHUB_ENV
```
Comment (edgarrmondragon) [inline .github/workflows/ci.yml:103]: At least until you want to actually publish wheels for 3.13

File: msgspec/_utils.py [hunk 2]
```diff
@@ -127,6 +127,7 @@ def get_class_annotations(obj):
                 value = type(None)
             elif isinstance(value, str):
                 value = _forward_ref(value)
+            # TODO: Pass `type_params` to `_eval_type` to fix deprecation warning
```
Comment (edgarrmondragon) [inline msgspec/_utils.py:130]: https://github.com/python/cpython/issues/118418

**Review by edgarrmondragon — COMMENTED** [review]
I think this is ready for an initial review.

Mostly followed advice from

* https://github.com/python/cpython/issues/121489#issuecomment-2213976007
* https://github.com/python/cpython/issues/121125#issuecomment-2213431408

**Comment by edgarrmondragon** [comment]
> This warning is currently raised but fixing it might be beyond the scope of this PR to fix it:
> 
> ```
> PytestUnraisableExceptionWarning: Exception ignored in PyDict_GetItem(); consider using PyDict_GetItemRef() or PyDict_GetItemWithError(): None
> ```

FWIW I think this is hiding an actual error:

```
PytestUnraisableExceptionWarning: Exception ignored in PyDict_GetItem(); consider using PyDict_GetItemRef() or PyDict_GetItemWithError(): None
    
    Traceback (most recent call last):
      File "/opt/_internal/cpython-3.13.0b3/lib/python3.13/typing.py", line 2251, in __hash__
        return hash((self.__origin__, self.__metadata__))
    TypeError: unhashable type: 'set'
    
      warnings.warn(pytest.PytestUnraisableExceptionWarning(msg))
```

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -99,10 +100,11 @@ jobs:
       - name: Set up Environment
         if: github.event_name != 'release'
         run: |
-            echo "CIBW_SKIP=${CIBW_SKIP} *-musllinux_* cp38-*_aarch64 cp39-*_aarch64 cp311-*_aarch64 cp312-*_aarch64" >> $GITHUB_ENV
+            echo "CIBW_BUILD=${CIBW_BUILD} cp313-*" >> $GITHUB_ENV
```
Comment (edgarrmondragon) [inline .github/workflows/ci.yml:103]: For the moment 3.13 wheels will only be built for non-release workflows.

**Comment by edgarrmondragon** [comment]
macos-11 runners aren't available anymore, so I updated it to macos-14 in 748f4bb.

https://github.blog/changelog/2024-05-20-actions-upcoming-changes-to-github-hosted-macos-runners/#macos-11-deprecation-and-removal

**Comment by edgarrmondragon** [comment]
> macos-11 runners aren't available anymore, so I updated it to macos-14 in [748f4bb](https://github.com/jcrist/msgspec/commit/748f4bb545c2489454bdafe1446d1c2bca9e8562).
> 
> https://github.blog/changelog/2024-05-20-actions-upcoming-changes-to-github-hosted-macos-runners/#macos-11-deprecation-and-removal

With that change, the wheels are finally built instead of the workflow getting stuck forever:

```
 12 wheels produced in 5 minutes:
  msgspec-0+untagged.1.g766a8f7-cp310-cp310-macosx_10_9_x86_64.whl    186 kB
  msgspec-0+untagged.1.g766a8f7-cp310-cp310-macosx_11_0_arm64.whl     180 kB
  msgspec-0+untagged.1.g766a8f7-cp311-cp311-macosx_10_9_x86_64.whl    184 kB
  msgspec-0+untagged.1.g766a8f7-cp311-cp311-macosx_11_0_arm64.whl     178 kB
  msgspec-0+untagged.1.g766a8f7-cp312-cp312-macosx_10_9_x86_64.whl    186 kB
  msgspec-0+untagged.1.g766a8f7-cp312-cp312-macosx_11_0_arm64.whl     180 kB
  msgspec-0+untagged.1.g766a8f7-cp313-cp313-macosx_10_13_x86_64.whl   186 kB
  msgspec-0+untagged.1.g766a8f7-cp313-cp313-macosx_11_0_arm64.whl     180 kB
  msgspec-0+untagged.1.g766a8f7-cp38-cp38-macosx_10_9_x86_64.whl      186 kB
  msgspec-0+untagged.1.g766a8f7-cp38-cp38-macosx_11_0_arm64.whl       180 kB
  msgspec-0+untagged.1.g766a8f7-cp39-cp39-macosx_10_9_x86_64.whl      186 kB
  msgspec-0+untagged.1.g766a8f7-cp39-cp39-macosx_11_0_arm64.whl       180 kB
```

https://github.com/edgarrmondragon/msgspec/actions/runs/10030128227/job/27719084757

**Comment by tacaswell** [comment]
I can also get msgspec to build with py313 with this branch, but not the branch in #703

**Comment by edgarrmondragon** [comment]
@jcrist This is ready for review at your convenience :)

File: .github/workflows/ci.yml (hunk 3 above)
Comment (edgarrmondragon) [reply .github/workflows/ci.yml]: ☝️ Outdated with the release of Python 3.13.0rc1 since we want to start publishing `cp313-*` wheels now.

**Comment by edgarrmondragon** [comment]
Updated to use Python 3.13.0rc1

**Comment by edgarrmondragon** [comment]
Updated to use Python 3.13.0rc2

**Comment by Tishka17** [comment]
Python 3.13 release is scheduled for upcoming Monday. Are there any plans to merge this PR and publish a new version?

**Comment by Tishka17** [comment]
Python 3.13 is out there.

**Comment by maurerle** [comment]
I do get a segfault though when importing msgspec.. Even with GIL enabled

```
(py313) ➜  msgspec git:(py313) ✗ python -X gil=1 -X faulthandler
Python 3.13.0 experimental free-threading build | packaged by conda-forge | (main, Oct  8 2024, 20:16:19) [GCC 13.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> 
>>> import msgspec
Fatal Python error: Segmentation fault

Current thread 0x00007fcd5e8ba640 (most recent call first):
  File "/home/maurer/gitea/msgspec/msgspec/__init__.py", line 19 in <module>
  File "<frozen importlib._bootstrap>", line 488 in _call_with_frames_removed
  File "<frozen importlib._bootstrap_external>", line 1022 in exec_module
  File "<frozen importlib._bootstrap>", line 935 in _load_unlocked
  File "<frozen importlib._bootstrap>", line 1331 in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 1360 in _find_and_load
  File "<python-input-1>", line 1 in <module>
```

**Comment by edgarrmondragon** [comment]
> I do get a segfault though when importing msgspec.. Even with GIL enabled
> 
> ```
> (py313) ➜  msgspec git:(py313) ✗ python -X gil=1 -X faulthandler
> Python 3.13.0 experimental free-threading build | packaged by conda-forge | (main, Oct  8 2024, 20:16:19) [GCC 13.3.0] on linux
> Type "help", "copyright", "credits" or "license" for more information.
> >>> 
> >>> import msgspec
> Fatal Python error: Segmentation fault
> 
> Current thread 0x00007fcd5e8ba640 (most recent call first):
>   File "~/github/msgspec/msgspec/__init__.py", line 19 in <module>
>   File "<frozen importlib._bootstrap>", line 488 in _call_with_frames_removed
>   File "<frozen importlib._bootstrap_external>", line 1022 in exec_module
>   File "<frozen importlib._bootstrap>", line 935 in _load_unlocked
>   File "<frozen importlib._bootstrap>", line 1331 in _find_and_load_unlocked
>   File "<frozen importlib._bootstrap>", line 1360 in _find_and_load
>   File "<python-input-1>", line 1 in <module>
> ```

@maurerle thanks for finding that. I'll give a read to https://py-free-threading.github.io/porting/, unless someone beats me to it and knows what's wrong here.

**Comment by maurerle** [comment]
When executing this on a gil build (e.g. `conda create -n py313-gil python=3.13`)
One gets an error if executed from the build directory:
```
>>> import msgspec
Traceback (most recent call last):
  File "<python-input-0>", line 1, in <module>
    import msgspec
  File "~/github/msgspec/msgspec/__init__.py", line 1, in <module>
    from ._core import (
    ...<14 lines>...
    )
ModuleNotFoundError: No module named 'msgspec._core'
```
So make sure to execute this not in the build directory :)
It works fine then :)
But I don't have a clue how to solve the above segfault either.

**Comment by jcrist** [comment]
Thanks for the PR (and everyone for their patience). With a new kid and recent layoff, I've been a bit busy and msgspec maintenance has languished. I really appreciate the help and patience here.

I have a few small fixups I want to land with this release, but I'd expect a release supporting Python 3.13 (not with freethreading) out in the next day or two. I'll work on adding freethreading support after getting that initial release out.

**Comment by ekneg54** [comment]
Hi,

with this merged I thought that I should be able to install `msgspec` on python 3.13. But it turns out, that this is not the case as you could see:
image

I wonder if I am missing something. Could you help?

**Comment by maurerle** [comment]
You are missing the python3-dev dependency.
Install it using your system package manager

**Comment by mj0nez** [comment]
It was merged but not released or am I mistaken? An install from main should work for sure [install from vcs](https://pip.pypa.io/en/stable/topics/vcs-support/)

**Comment by ekneg54** [comment]
thank you. I wait for the release. Do you know when it is planned?

**Comment by mj0nez** [comment]
No, but soon I guess.

**Comment by st3fan** [comment]
@jcrist Is there anything the community can do to help to get a release out? There is probably some collective spare hacking time available during the break now. Let us know if you need help testing or writing release notes, etc.

**Pull request merged**